zipline/src/engineTest/kotlin/app/cash/zipline/SourceMapParserTest.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (C) 2021 Square, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package app.cash.zipline
+
+import app.cash.zipline.internal.SourceMapParser
+import app.cash.zipline.internal.readVarint
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertTrue
+import okio.Buffer
+import okio.FileSystem
+import okio.Path.Companion.toPath
+
+class SourceMapParserTest {
+  @Test fun simpleSourceMap() {
+    val sourceMapJson = """
+      {
+        "version": 3,
+        "file": "kotlin-js-demo.js",
+        "sources": [
+          "../../../../../src/main/kotlin/main.kt"
+        ],
+        "sourcesContent": [
+          null
+        ],
+        "names": [],
+        "mappings": ";;;;;;;;;;;;EAAA,gB;IACE,YAAY,c;IACZ,OAAQ,KAAI,KAAJ,C;EACV,C;;;;;;"
+      }
+      """.trimIndent()
+    val element = Element(
+      line = "<anonymous>",
+      fileName = "demo.js",
+      lineNumber = 14,
+    )
+
+    assertEquals(
+      """
+      at <anonymous> ../../../../../src/main/kotlin/main.kt:2
+      """.trimIndent(),
+      decodeStackTrace(
+        stackTrace = listOf(element),
+        sourceMapJsonLoader = { sourceMapJson },
+      )
+    )
+  }
+
+  @Test fun readVarints() {
+    val buffer = Buffer().writeUtf8("IAkpZI")
+    assertEquals(4L, buffer.readVarint())
+    assertEquals(0L, buffer.readVarint())
+    assertEquals(12946L, buffer.readVarint())
+    assertEquals(4L, buffer.readVarint())
+    assertTrue(buffer.exhausted())
+  }
+
+  @Test fun mapStacktrace() {
```

Comment:
This doesn't seem like it should be / was meant to be committed.

Reply:
It'd be useful to have this as an end-to-end test with more complex input than `simpleSourceMap`, but looks like we'd need to copy over a bunch of map files as they're not available on all platforms. Or do you think `simpleSourceMap` with a few more cases, such as optional name, is enough?

Reply:
I agree a more comprehensive test is needed but this reads arbitrary data from the filesystem without appropriate task dependencies. The source files may not be there, or may be out-of-date.

Aside from that, there's a lot of logic that's going on here tangential to the API under test. There's entirely new input types with a mapping and rendering function. The API under test is hidden under all this.

I would expect an end-to-end test to be more focused on our eventual usage which is remapping bytecode.

Reply:
Sounds good, deleted 25d9fab